### ISSUE TITLE
JsonFormat fails to parse Json with both lowerCamelCase and snake_case fields

### ISSUE THREAD

**Issue by yuri-sergiichuk** [protocolbuffers/protobuf#7641]
**What version of protobuf and what language are you using?**
Version: v3.12.2
Language: Java

**What operating system (Linux, Windows, ...) and version?**
Windows 10

**What runtime / compiler are you using (e.g., python version or gcc version)**
openjdk version "11.0.7" 2020-04-14 LTS
OpenJDK Runtime Environment Corretto-11.0.7.10.1 (build 11.0.7+10-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.7.10.1 (build 11.0.7+10-LTS, mixed mode)

**What did you do?**

Try to deserialize a proto message from JSON where the message field appears in both `lowerCamelCase` and `snake_case` notations, e.g. `message_id` and `messageId`. An attempt to deserialize such a message leads to `InvalidProtocolBufferException`.

Here is a companion repository for the issue: https://github.com/yuri-sergiichuk/proto-json-same-fields-issue

**What did you expect to see**

The message is deserialized properly taking into account the field that has higher priority (probably, the `lowerCamelCase` one as it is the default notation for JSON proto schema).

**What did you see instead?**

An `InvalidProtocolBufferException` exception being thrown:

```
Exception in thread "main" java.lang.IllegalArgumentException: {
  "message": {
    "data": "eyJrZXkiOiJ2YWx1ZSJ9",
    "messageId": "450292511223766",
    "message_id": "450292511223766",
    "publishTime": "2020-06-21T20:48:25.908Z",
    "publish_time": "2020-06-21T20:48:25.908Z"
  },
  "subscription": "projects/test-project/subscriptions/test-subscription"
} cannot be parsed to the class com.google.pubsub.v1.PubsubPushNotification class.
	at io.spine.util.Exceptions.newIllegalArgumentException(Exceptions.java:181)
	at io.spine.json.Json.fromJson(Json.java:117)
	at com.github.proto.TestDeserialization.main(TestDeserialization.java:10)
Caused by: com.google.protobuf.InvalidProtocolBufferException: Field google.pubsub.v1.PubsubMessage.message_id has already been set.
	at com.google.protobuf.util.JsonFormat$ParserImpl.mergeField(JsonFormat.java:1650)
Caused by: com.google.protobuf.InvalidProtocolBufferException: Field google.pubsub.v1.PubsubMessage.message_id has already been set.

	at com.google.protobuf.util.JsonFormat$ParserImpl.mergeMessage(JsonFormat.java:1502)
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1460)
	at com.google.protobuf.util.JsonFormat$ParserImpl.parseFieldValue(JsonFormat.java:2001)
	at com.google.protobuf.util.JsonFormat$ParserImpl.mergeField(JsonFormat.java:1665)
	at com.google.protobuf.util.JsonFormat$ParserImpl.mergeMessage(JsonFormat.java:1502)
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1460)
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1342)
	at com.google.protobuf.util.JsonFormat$Parser.merge(JsonFormat.java:477)
```

**Test case**

I've added a straightforward test case to show better the scenario we're facing: https://github.com/yuri-sergiichuk/proto-json-same-fields-issue/blob/master/src/test/java/com/github/proto/ProtoDeserializationTest.java

**Comment by anandolee** [comment]
We are working on enforce style guide. Field name like "messageId' does not match our style guide.
For internal google usages, we will raise errors at compile time if the .proto file is not match our style guide. For open source, we raise warnings but may change to raise errors too sometime in future.

The first step for enforce style guide will include:
-Enum names must always be UPPER_CASE
-Field names must always be lower_case
-Message names must always be CamelCase

**Comment by wd40bomber7** [comment]
Hey @anandolee , I'm the owner of this [two year old issue](https://github.com/protocolbuffers/protobuf/issues/5414 ) which also boils down to us not using "snake case" and running into trouble.

I'm just incredibly confused why so many years into the project protobuf would decide to start enforcing a specific style? Even right this very second there are zero mentions of "snake case" [in the proto 3 documentation](https://developers.google.com/protocol-buffers/docs/proto3). How could you think of starting to warn/error on style without even mentioning the recommended style to users?

We have a significant quantity of code that uses protobuf with PascalCase fields and it would be disastrously expensive to rewrite all of it to interact with snake_case fields. Not to mention some of the services we talk to are 3rd party (JSON) and we have no control to enforce that they use snake_case fields. We're going to be stuck with whatever version of protobuf is the last one before snake_case becomes enforced, why not introduce proto 4 and make snake case a breaking change/feature of that language from the start?

This seems absolutely misguided to me. It might make sense internal to google where there is a homogoneous style that has been internally enforced this whole time but it makes no sense in the public project where anyone could have done anything for _years_ up to this point.

I would love to discuss this further with stake holders, is there a good place for that?

**Comment by yuri-sergiichuk** [comment]
Hey, @anandolee thank you for looking at it. I do agree with the approach of leading developers into some canonical direction, but as @wd40bomber7 mentions, it is not always possible to force a particular style.
Moreover, our issue here is that the Protobuf definition in the example repository has only `snake_case` fields.
Here is our test Proto wrapper around Google's PubSub message:
```
// Pubsub push message notification.
message PubsubPushNotification {

    // Pubsub message.
    PubsubMessage message = 1 [(required) = true];

    // The name of the Pubsub subscription that pushed the current notification.
    // Format is `projects/{project}/subscriptions/{subscription}`.
    //
    string subscription = 2 [(required) = true, (pattern).regex = "projects/.+/subscriptions/.+"];
}
```
And here is the PubSub message itself (the one that we have issues with now):
```
message PubsubMessage {
  // The message data field. If this field is empty, the message must contain
  // at least one attribute.
  bytes data = 1;

  // Attributes for this message. If this field is empty, the message must
  // contain non-empty data.
  map<string, string> attributes = 2;

  // ID of this message, assigned by the server when the message is published.
  // Guaranteed to be unique within the topic. This value may be read by a
  // subscriber that receives a `PubsubMessage` via a `Pull` call or a push
  // delivery. It must not be populated by the publisher in a `Publish` call.
  string message_id = 3;

  // The time at which the message was published, populated by the server when
  // it receives the `Publish` call. It must not be populated by the
  // publisher in a `Publish` call.
  google.protobuf.Timestamp publish_time = 4;

  // If non-empty, identifies related messages for which publish order should be
  // respected. If a `Subscription` has `enable_message_ordering` set to `true`,
  // messages published with the same non-empty `ordering_key` value will be
  // delivered to subscribers in the order in which they are received by the
  // Pub/Sub system. All `PubsubMessage`s published in a given `PublishRequest`
  // must specify the same `ordering_key` value.
  // <b>EXPERIMENTAL:</b> This feature is part of a closed alpha release. This
  // API might be changed in backward-incompatible ways and is not recommended
  // for production use. It is not subject to any SLA or deprecation policy.
  string ordering_key = 5;
}
```

The bigger (from my perspective) issue is that with 3rd parties, you never control the actual JSON they're sending. And in this particular case, for the defined `PubsubMessage` proto we actually receive smth like this:
```
{
  "message": {
    "data": "eyJrZXkiOiJ2YWx1ZSJ9",
    "messageId": "450292511223766",
    "message_id": "450292511223766",
    "publishTime": "2020-06-21T20:48:25.908Z",
    "publish_time": "2020-06-21T20:48:25.908Z"
  },
  "subscription": "projects/test-project/subscriptions/test-subscription"
}
```

So even with proper proto definitions, the parsing still fails. I'd be also glad to discuss the approach the Protobuffers team would like to choose here. We rely on Proto heavily and just love the idea that when the 3rd party does not provide their own SDK or an idiomatic way to use the API, we can simply create protos and consume their JSON API in protos.

**Comment by yuri-sergiichuk** [comment]
@anandolee I've added a straightforward test case that highlights the problem here: https://github.com/yuri-sergiichuk/proto-json-same-fields-issue/blob/master/src/test/java/com/github/proto/ProtoDeserializationTest.java

**Comment by mcy** [comment]
tl;dr snake/camel case distinction is a red herring; the issue is that optional fields must not be specified more than once.

Unfortunately we don't have a public spec for all of JSON (although we are working on fixing it...).

JSON supports parsing either via:
- The name declared in the .proto file.
- The name declated in the `json_name` option, which defaults to the camelcasification of the name declared in the .proto.

However, the issue at hand is not that: it's that you're specifying a name multiple times in JSON. Writing

```json
{
  "my_field": 1,
  "myField": 2,
}
```

is equivalent to

```json
{
  "my_field": 1,
  "my_field": 2,
}
```

which is not valid JSON because a singular field is set multiple times.

**Comment by parwalayush** [comment]
@mcy 
Even after adding json_name tag, the java generated file consists of camel cases and not snake_case.
Is this an expected behavior?
  private SendSmsRequestt() {
    phoneNumber_ = "";
  }
  Anyway to have a snake_case?
